### ISSUE TITLE
Create Sitemap for Better SEO

### ISSUE THREAD

**Issue by MohammadPCh** [Giveth/giveth-dapps-v2#915]
We need to generate a sitemap for our website to search engines can find our pages better and easier. now they can only crawl pages linked to our popular pages (homepage or first load of projects).
I suggest creating the sitemap once per day. We need the devOps team's help with this.
@aminlatifi @MoeNick

**Comment by MohammadPCh** [comment]
This google doc for Sitemap:
https://developers.google.com/search/docs/advanced/sitemaps/overview

**Comment by divine-comedian** [comment]
@mhmdksh @MohammadPCh - is this issue still relevant? 

I'm VERY interested in improving our SEO.

**Comment by divine-comedian** [comment]
@mhmdksh can you please provide an update on this issue?

**Comment by mhmdksh** [comment]
> @mhmdksh can you please provide an update on this issue?

@divine-comedian Sorry I can't provide any update on this neither this is my field of expertise. I prefer if I get unassigned as this is a Frontend/SEO-Optimization issue, not Devops work.

Thanks

@MohammadPCh Anyone can take this??

**Comment by divine-comedian** [comment]
@MohammadPCh can this issue be passed to someone or what is the next course of action?

**Comment by MohammadPCh** [comment]
@kkatusic @lovelgeorge99 @Reshzera guys who can take it?
we need a script to get projects from the server, add the static pages, and create a sitemap.
let me know if you need help

**Comment by kkatusic** [comment]
> @kkatusic @lovelgeorge99 @Reshzera Hey everyone, who has the bandwidth to take this on?
> 
> We need a script to:
> 
> 1. Retrieve projects, users, and campaigns from the server
> 2. Add the static pages
> 3. Create a sitemap
> 
> Let me know if you need any help.

I will take it, thx @MohammadPCh

**Comment by kkatusic** [comment]
@divine-comedian and @MohammadPCh, cc @CarlosQ96  is it ok to expose users inside sitemap?

**Comment by divine-comedian** [comment]
I am not sure about what the implications are so I would defer to what either of them think

**Comment by kkatusic** [comment]
Let's than wait @MohammadPCh and @CarlosQ96 responsne ;)

**Comment by MohammadPCh** [comment]
@kkatusic Yes we can add the public profile to the sitemap. but we should disallow the dashboard and private zones.

**Comment by CarlosQ96** [comment]
Ok we should make a query just to expose the minimum required info of users.
Also how often will the sitemap be generated? Ideally we could have a ipfs updated every day. I'm not sure a open public endpoint is that efficient even with cache. It's loading too much data into memory.

**Comment by CarlosQ96** [comment]
@kkatusic @MohammadPCh

**Comment by CarlosQ96** [comment]
@kkatusic Sounds good to make the cronjob once per week. Exposing public data only.

**Comment by divine-comedian** [comment]
need to provide some back-end function for listing users, need to generate user sitemap only with public data on front-end.

@kkatusic please review and add estimate of when this can be completed.

**Comment by kkatusic** [comment]
> need to provide some back-end function for listing users, need to generate user sitemap only with public data on front-end.
> 
> @kkatusic please review and add estimate of when this can be completed.

I will try to finish this week, but have now problem with BE, we will try to fix that before because I need BE functions also to finish this one.

**Comment by divine-comedian** [comment]
@CarlosQ96 can you give #1887 a review?

**Comment by divine-comedian** [comment]
Needs PR Review by @CarlosQ96

**Comment by kkatusic** [comment]
@divine-comedian, cc @CarlosQ96 , files generated fine, but project contain some character and I improved function for escaping that characters, sitemap files:

- main one that include all: https://staging.giveth.io/sitemap.xml
- projects, this has old error: https://staging.giveth.io/sitemap/projects-sitemap.xml
- qf round: https://staging.giveth.io/sitemap/qf-sitemap.xml
- users: https://staging.giveth.io/sitemap/users-sitemap.xml

**Comment by divine-comedian** [comment]
Thanks, so what are the next steps @kkatusic - do we need to wait again until next Sunday to see the results?

**Comment by kkatusic** [comment]
> Thanks, so what are the next steps @kkatusic - do we need to wait again until next Sunday to see the results?

I will check again on next Monday. I generated on locahost and it's fine, but you never know what new character can appears.

**Comment by MoeNick** [comment]
@kkatusic is on it

**Comment by divine-comedian** [comment]
bump on this @CarlosQ96

**Comment by kkatusic** [comment]
@divine-comedian, cc @CarlosQ96, FE is merged, we will check BE soon

**Comment by divine-comedian** [comment]
@CarlosQ96 @kkatusic this PR still needs a merge

**Comment by divine-comedian** [comment]
need to resolve https://github.com/Giveth/impact-graph/issues/1922 before we can safely merge the related PR

**Comment by divine-comedian** [comment]
@kkatusic will try to run this job tomorrow and make sure results are what we expect

**Comment by kkatusic** [comment]
@divine-comedian again I stuck with this, after update on BE side it was resolving wrong IP address, for some reason hosts file has been set up for looking same FE url on same server, Moe helped and removed that after he found problem. 
Then new issue, Verel don't allow you to change file or add new file inside public folder or any other problem, files are only read-only state. 
Then I tried to update files to Pinata and have draft working copy, by Pinata is changing every time uploaded file URL, this not acceptable for us, I think we will need to move this URL data to database, then I need more time, migration, database table, resolver, api url....

**Comment by kkatusic** [comment]
How to test it, on staging link to sitemap would be:

https://staging.giveth.io/sitemap.xml

at the end of the file you should see 3 more links to another sitemaps:

![Image](https://github.com/user-attachments/assets/ed3b2f06-c9a2-4d7b-97ec-f2eeed26e04c)

and you will check that url. On production it will run on Sunday after midnight

**Comment by divine-comedian** [comment]
Thanks! @maryjaf are you free to test this? 

@kkatusic is there anything else we should verify on this issue? can we assume the content of the ipfs files are correct? 

Is there anything we could confirm regarding SEO rating at this point or only when it is on production?

**Comment by kkatusic** [comment]
@divine-comedian, cc @maryjaf, maybe randomly pick up one or more projects or usera or QF rounds link from the sitemap list and check if it exist.

When this sitemaps generate on production you only submit main sitemap file: https://giveth.io/sitemap.xml I think inside google search console and it will parse it.

SEO meta tags are there inside each main link and it's working fine, fo other stuff I don't know, there is always something to improve in SEO.

**Comment by maryjaf** [comment]
I’m not sure how can I test this issue, could you please help me on this? 
Is qa test required for this issue?

**Comment by kkatusic** [comment]
> I’m not sure how can I test this issue, could you please help me on this Is qa test required for this issue?

It's list of the all links that Giveth have, also inside main list there is link to Pinata file where is XML file for all projects links, users public data and QF rounds. If you open current link for example inside :

https://giveth.mypinata.cloud/ipfs/QmThQhddXtRJRR7sfnyGjgrkxdBNNKQgLYZHh2eh8mj8ww

you see list of the links of the all project on staging:

![Image](https://github.com/user-attachments/assets/67285818-70b4-4ae8-9758-5d8c855cd868)

and if you choose one link:

https://staging.giveth.io/project/Lillian-Bay-Foundation

it should open project, not some cancelled or deleted project.

More about XML Sitemap file you can find here:

https://www.elegantthemes.com/blog/wordpress/xml-sitemap-seo

**Comment by divine-comedian** [comment]
@kkatusic - I have review the generated files and based on the article you shared I have some feedback on WHICH files we should generate 

1. we should remove these pages: 
- claim (GIVdrop ended 2 years ago)
- create (user needs a crypto wallet and to be signed-in to access this page)

2. we should add these pages: 
- category projects pages (art & culture, community, etc..)

3. about page shows up twice in generated file, we should remove the duplicate

4. include only verified projects in generated IPFS files to ensure indexed pages are higher quality

5. how does sitemap and SEO handle translated pages such as spanish pages of Giveth?

**Comment by kkatusic** [comment]
@divine-comedian I fixed all that you reported except translated pages. For titles and description of the translated pages we are using same content: https://github.com/Giveth/giveth-dapps-v2/blob/develop/src/content/metatags.ts , I will need then add similar for all available languages. Also then I need to generate same files for all projects, rounds and user using middle suffix for domain, for example: [EN] giveth.io/test-project => [ES] giveth.io/es/test-project => [CT] giveth.io/es/test-project, but I don't know do we need this for same, because titles and description of the project are same for each language, don't know how Google will order/mark that sitemap results

**Comment by divine-comedian** [comment]
@maryjaf  - can you test this on sunday - the script should run at midnight on saturday/sunday

**Comment by maryjaf** [comment]
1. I haven't seen this in XML file: @kkatusic 

- [ ] homepage
- [ ] all projects
- [ ] qf all
- [ ] qf archive
- [ ] category pages
- [ ] onboarding/donors
- [ ] onboarding/giveconomy

_________________________

> in addition there should be some ipfs files that contain project data
> * [ ]  ONLY VERIFIED/VOUCHED projects should be included in ipfs files.

could you please explain more about this item, I see the below context in XML file @divine-comedian 

<loc>https://giveth.mypinata.cloud/ipfs/QmRcMHxZVDbmoXFXqWwFPwToQGdnEGAsutyCbod5JT1PTo</loc>
<lastmod>2025-03-09T09:44:36.175Z</lastmod>
</url>
<url>
<loc>https://giveth.mypinata.cloud/ipfs/QmUtCH85VCnxnzSUfrMfcrNNTK2w8enTDE8ivhG51rK4Xa</loc>
<lastmod>2025-03-09T09:44:36.175Z</lastmod>
</url>
<url>
<loc>https://giveth.mypinata.cloud/ipfs/QmdjXkdt83k2TzPrpMrwavD9GSkRsw1ZhCnZe1Pffm83XX</loc>
<lastmod>2025-03-09T09:44:36.175Z</lastmod>
</url>

**Comment by kkatusic** [comment]
@maryjaf can you check it now, thx

https://staging.giveth.io/sitemap.xml

**Comment by maryjaf** [comment]
> 1. I haven't seen this in XML file: [@kkatusic](https://github.com/kkatusic)
> 
> * [x]  homepage[x]  qf all[x]  category pages
> 
> I see **donors/giveconomy** and **donors/projects** instead of below items
> 
> * [x]  onboarding/donors[x]  onboarding/giveconomy

Thanks @kkatusic it has been fixed

**Comment by maryjaf** [comment]
> in addition there should be some ipfs files that contain project data
> 
> * [ ]   ONLY VERIFIED/VOUCHED projects should be included in ipfs files.

but about this item for example I see this project https://giveth.io/project/my-ptoject in ips file and it isn't verified

**Comment by maryjaf** [comment]
> in addition there should be some ipfs files that contain project data
> 
> * [ ]   ONLY VERIFIED/VOUCHED projects should be included in ipfs files.

but about this item I see [test-project0001-01](https://giveth.io/project/test-project0001-01) project and it's status isn't Verified

![Image](https://github.com/user-attachments/assets/a673e051-9040-40a1-a989-cc088920f33a)

**Comment by kkatusic** [comment]
@maryjaf I can't find that project inside file, can you provide me in what ipfs file you are looking? thx

**Comment by maryjaf** [comment]
> [@maryjaf](https://github.com/maryjaf) I can't find that project inside file, can you provide me in what ipfs file you are looking? thx

https://giveth.mypinata.cloud/ipfs/QmRcMHxZVDbmoXFXqWwFPwToQGdnEGAsutyCbod5JT1PTo

**Comment by maryjaf** [comment]
Thanks @kkatusic I've checked again and all projects status is Verified 
We can call it Done

**Comment by kkatusic** [comment]
thx @maryjaf
